### PR TITLE
Add clear and copy methods to TraitListObject

### LINKED ISSUE #679: TraitListObject should grow copy and clear methods
This would allow some unification of paths between List and Set/Dict in a few places.

### PR DESCRIPTION
**Pull request by shoeb-github**
closes #679

### CODE REVIEW

File: traits/trait_list_object.py [hunk 1]
```diff
@@ -416,6 +416,23 @@ def rename(self, name):
...
+
+        removed = list.copy(self)
+        list.clear(self)
+        index = 0
+
+        self._send_trait_items_event(
+            self.name_items, TraitListEvent(index, removed)
+        )
+
+    def copy(self):
```
Comment (mdickinson) [inline traits/trait_list_object.py]: So in discussion with @corranwebster, we realised that actually there's no need for a `copy` method at all; we should just inherit the `list.copy` base class method.

In particular, `copy` on a `List` value should return a plain old Python list.

Suggest removing this method, but keeping the tests, and adding an additional test to verify that the result of the `copy` is a `list`, not a `TraitListObject`.

File: traits/trait_list_object.py [hunk 2]
```diff
@@ -416,6 +416,23 @@ def rename(self, name):
...
 
+    def clear(self):
+        if self.trait.minlen > 0:
+            self.len_error(0)
+
+        removed = list.copy(self)
+        list.clear(self)
+        index = 0
+
+        self._send_trait_items_event(
```
Comment (mdickinson) [inline traits/trait_list_object.py]: As discussed in the meeting, in the case of an empty list there shouldn't be any items changed event. Please could you add a test for that corner case?

File: traits/trait_list_object.py (hunk 1 above)
Comment (shoeb-github) [reply traits/trait_list_object.py]: Thanks, I removed the `copy` method and added the test.

File: traits/trait_list_object.py (hunk 2 above)
Comment (shoeb-github) [reply traits/trait_list_object.py]: I replaced the this with `del self[:]` but that causes a items changed notification when an empty list is cleared and thus, a test failure. Is this behavior for `del [:]` expected?

Comment (mdickinson) [reply traits/trait_list_object.py]: Okay, that's interesting. :-) I'd expect `del mylist[:]` *not* to create an event if the list is already empty (and similarly, if `mylist` is a list of length 2, I'd expect something like `del mylist[2:]` not to create an event).

Let's keep this PR focused on `clear` and `copy`, but please could you open a bug report for the behaviour of `del` here?

Comment (mdickinson) [reply traits/trait_list_object.py]: ... and now I'm wondering what (for example) `mylist.extend([])` does. But again, that's off-topic for this PR.

File: traits/trait_list_object.py [hunk 3]
```diff
@@ -416,6 +416,9 @@ def rename(self, name):
             self.name = name
             self.trait = trait.handler
 
+    def clear(self):
+        del self[:]
```
Comment (mdickinson) [inline traits/trait_list_object.py]: I think we should still go through `list.clear` here, painful though that may be; just on general principles, each of the `TraitListObject` standard list methods should call the underlying list method. I can't *think* offhand of a case where that would matter for `clear`, but I'd prefer to be safe.

File: traits/trait_list_object.py (hunk 2 above)
Comment (mdickinson) [reply traits/trait_list_object.py]: > please could you open a bug report for the behaviour of `del` here

Sorry, I forgot that you aren't around for the next couple of days; I've opened #738 for this.

Comment (mdickinson) [reply traits/trait_list_object.py]: And for the record, `.extend([])` seems to be fine; it doesn't fire an event.

**Comment by shoeb-github** [comment]
Thanks for opening the issue. Reverted back to `list.clear` implementation.

File: traits/trait_list_object.py [hunk 4]
```diff
@@ -416,6 +416,21 @@ def rename(self, name):
             self.name = name
             self.trait = trait.handler
 
+    def clear(self):
+        if self.trait.minlen > 0:
+            self.len_error(0)
+
+        if len(self) == 0:
```
Comment (mdickinson) [inline traits/trait_list_object.py]: Can we restructure slightly so that `list.clear` is still called when `len(self) == 0`? Right now `list.clear` on an empty list is almost certainly a no-op, but we can't know for sure that Python won't add some functionality at some point in the future.

File: traits/trait_list_object.py [hunk 5]
```diff
@@ -416,6 +416,21 @@ def rename(self, name):
...
+            return
+
+        removed = list.copy(self)
```
Comment (mdickinson) [inline traits/trait_list_object.py]: `self.copy()` might look a bit more natural here. But this is likely fine, too.

Comment (shoeb-github) [reply traits/trait_list_object.py]: Replaced with `self.copy()` and fixed to call `list.clear` for empty lists too in 2bcd7b4.
